File: chapitre5/exercice4.c
#include <pthread.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <unistd.h> 

typedef struct {
    int n;
    int *t;
} TypeTableau;

typedef struct {
    TypeTableau t1;
    TypeTableau t2;
} ParamInclusion;

void *generation(void *arg) {
    TypeTableau *tab = (TypeTableau *)arg;
    printf("Génération d'un tableau de %d éléments...\n", tab->n);
    for(int i = 0; i < tab->n; i++) {
        tab->t[i] = rand() % 100; 
    }
    sleep(1); 
    pthread_exit(NULL);
}

void* t_inclusion(void* arg) {
    ParamInclusion* p = (ParamInclusion*)arg;
    long est_inclus = 1; 

    for (int i = 0; i < p->t1.n; i++) {
        int trouve = 0;
        for (int j = 0; j < p->t2.n; j++) {
            if (p->t1.t[i] == p->t2.t[j]) {
                trouve = 1;
                break; 
            }
        }
        if (trouve == 0) {
            est_inclus = 0; 
            break;
        }
    }
    pthread_exit((void*)est_inclus); 
}

void* t_annulation(void* arg) {
    char c;
    while (1) {
        c = getchar();
        if (c == 'A' || c == 'a') {
            printf("\n[ANNULATION] Arrêt du programme par l'utilisateur.\n");
            exit(0); 
        }
    }
    return NULL;
}

int main() {
    pthread_t t1, t2, recherche, annulation;
    TypeTableau data1, data2;
    int n1, n2;
    void* status; 

    srand(time(NULL));

    do {
        printf("Taille T1 et T2 (T1 doit être <= T2) : ");
        scanf("%d %d", &n1, &n2);
    } while (n1 > n2);
    
    // Nettoyage du tampon pour le futur getchar()
    while (getchar() != '\n'); 

    data1.n = n1;
    data1.t = malloc(n1 * sizeof(int));
    data2.n = n2;
    data2.t = malloc(n2 * sizeof(int));

    pthread_create(&t1, NULL, generation, (void*)&data1); 
    pthread_create(&t2, NULL, generation, (void*)&data2); 
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    ParamInclusion total = {data1, data2};

    pthread_create(&recherche, NULL, t_inclusion, (void*)&total);
    pthread_create(&annulation, NULL, t_annulation, NULL);

    pthread_join(recherche, &status);

    printf("Résultat du calcul : %s\n", (long)status ? "Inclus" : "Non inclus");

    // Libération de la mémoire [4]
    free(data1.t);
    free(data2.t);
   
    return 0;
}

File: chapitre5/exercice4.py
import threading
import random
import time
import sys

def generation(tab):
    print(f"Génération d'un tableau de {tab['n']} éléments...")
    tab['t'] = [random.randint(0, 99) for _ in range(tab['n'])]
    time.sleep(1)

def t_inclusion(p, result):
    est_inclus = True
    for i in range(p['t1']['n']):
        trouve = False
        for j in range(p['t2']['n']):
            if p['t1']['t'][i] == p['t2']['t'][j]:
                trouve = True
                break
        if not trouve:
            est_inclus = False
            break
    result['value'] = est_inclus

def t_annulation(stop_event):
    while not stop_event.is_set():
        c = sys.stdin.read(1)
        if c.lower() == 'a':
            print("\n[ANNULATION] Arrêt du programme par l'utilisateur.")
            sys.exit(0)

def main():
    random.seed()

    while True:
        try:
            entry = input("Taille T1 et T2 (T1 doit être <= T2) : ")
            n1, n2 = map(int, entry.split())
            if n1 <= n2:
                break
        except ValueError:
            pass

    data1 = {'n': n1, 't': []}
    data2 = {'n': n2, 't': []}

    thread1 = threading.Thread(target=generation, args=(data1,))
    thread2 = threading.Thread(target=generation, args=(data2,))
    thread1.start()
    thread2.start()
    thread1.join()
    thread2.join()

    total = {
        't1': data1,
        't2': data2
    }

    result = {'value': False}
    stop_event = threading.Event()

    recherche = threading.Thread(target=t_inclusion, args=(total, result))
    annulation = threading.Thread(target=t_annulation, args=(stop_event,), daemon=True)

    recherche.start()
    annulation.start()

    recherche.join()
    stop_event.set()

    print(f"Résultat du calcul : {'Inclus' if result['value'] else 'Non inclus'}")

if __name__ == "__main__":
    main()